include/fast_float/float_common.h

```diff
@@ -111,15 +111,31 @@ fastfloat_really_inline value128 full_multiplication(uint64_t value1,
   return answer;
 }
 
-#else
+#else // gcc
 
 // compute value1 * value2
-fastfloat_really_inline value128 full_multiplication(uint64_t value1,
-                                                     uint64_t value2) {
+fastfloat_really_inline value128 full_multiplication(uint64_t a,
+                                                     uint64_t b) {
   value128 answer;
-  __uint128_t r = ((__uint128_t)value1) * value2;
+#if  defined(__i386) || defined(__i386__) || defined(_M_IX86) || defined(__arm__)
+  static constexpr const uint64_t lo32 = 0xffffffffu;
+  // https://stackoverflow.com/questions/28868367/getting-the-high-part-of-64-bit-integer-multiplication
+  uint64_t a_lo = a & lo32;
+  uint64_t a_hi = a >> 32;
+  uint64_t b_lo = b & lo32;
+  uint64_t b_hi = b >> 32;
+  uint64_t ab_hi =  a_hi * b_hi;
+  uint64_t ab_mid = a_hi * b_lo;
+  uint64_t ba_mid = b_hi * a_lo;
+  uint64_t ab_lo =  a_lo * b_lo;
+  uint64_t carry_bit = ((ab_mid & lo32) + (ba_mid & lo32) + (ab_lo >> 32)) >> 32;
+  answer.high = ab_hi + (ab_mid >> 32) + (ba_mid >> 32) + carry_bit;
+  answer.low = ab_lo + (ab_mid & lo32) + (ba_mid & lo32);
```

Comment:
not sure about this. @lemire can you confirm?

Reply:
It looks good at a glance. And it is not obviously inefficient. If the logic is somehow wrong (could happen, there are many variables here), then a test would almost surely reveal the mistake. It is hard to be "almost correct" with these problems. :-)

Reply:
Turns out you already had the working code there :-). Just needed to shift around the feature test macros.